proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,209 @@
+`hlsl_intrinsics.h`:
+
+```c++
+__attribute__((availability(shadermodel, introduced = 6.0)))
+__attribute__((clang_builtin_alias(__builtin_hlsl_wave_active_count_bits)))
+uint WaveActiveCountBits(bool Bit);
+```
+
+> Note: the actual header uses macros to condense the attribute descriptions.
+> The example expands the macros for explicitness.
+
+The example above declares the `WaveActiveCountBits` function to require shader
+model 6.0+. This AST annotation allows clangd to communicate to users the
+availability information to language server protocol clients. This can drive
+more informed auto-complete, refactoring, and in-editor diagnostics.
+
+### Diagnostic Modes
+
+This proposal introduces three new modes for diagnosing API availability:
+relaxed, strict, and pedantic. Clang's default mode will be the
+strict mode. The `-fhlsl-diagnostic-mode=` flag allows users to select which
+diagnostic mode they wish to use.
+
+### Relaxed Diagnostic Mode
+
+The relaxed diagnostic mode performs an AST traversal after the translation unit
+has been fully parsed, and requires construction of a call graph. In the relaxed
+mode, an AST visitor will traverse to all `CallExpr` nodes that are reachable
+from exported functions (either library exports or entry functions). If the
+callee of a `CallExpr` has availability annotations that signify that the API is
+unavailable for the target shader model and stage the compiler emits a
+_warning_.
+
+Clang encodes the target shader model version in the target triple, and the
+shader stage in the `HLSLShaderAttr` which is implicitly or explicitly applied
+to the entry function.
+
+The relaxed mode does not issue diagnostics for `CallExpr` nodes that are inside
+functions which are not reachable from exported functions.
+
+### Strict Diagnostic Mode
+
+The implementation of the strict diagnostic mode matches the relaxed mode,
+except that when a `CallExpr` references an unavailable API, the compiler emits
+an _error_.
+
+The strict mode does not issue diagnostics for `CallExpr` nodes that are inside
+functions which are not reachable from exported functions.
+
+### Pedantic Diagnostic Mode
+
+The pedantic diagnostic mode strives to aggressively issue diagnostics. For
+non-library shaders, during parsing any callee of a `CallExpr` that has
+availability annotation that marks it as unavailable for the target shader model
+and stage will produce an _error_.
+
+For library shaders, during parsing any callee of a `CallExpr` that has
+availability annotation that marks it as unavailable for the target shader model
+version will produce an _error_. After parsing the translation unit, an AST
+visitor will traverse to all `CallExpr` nodes that are reachable from annotated
+shader entry functions. If the callee of a `CallExpr` has availability
+annotations that signify that the API is unavailable for the target shader stage
+the compiler emits an _error_.

```

Comment:
Are strict and pedantic at all different for library shaders? The explanations sound the same.

Reply:
Yes, but I can elaborate.